new permutedims without unrolling permutation orders

GiggleLiu opened the pull request:
fix #375 
## Benchmarks
1. low rank arrays

OLD:
```julia
julia> using CUDA

julia> t = CUDA.randn(200, 200, 200);

julia> using BenchmarkTools

julia> @benchmark CUDA.@sync permutedims($t, (3,2,1))
BenchmarkTools.Trial: 1572 samples with 1 evaluation.
 Range (min … max):  1.624 ms … 514.483 ms  ┊ GC (min … max): 0.00% … 0.49%
 Time  (median):     1.917 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.167 ms ±  24.906 ms  ┊ GC (mean ± σ):  0.16% ± 0.02%

                                         ▁▂▄▅▆██▆▅▄▃▃▁         
  ▂▂▂▂▂▃▃▃▂▃▃▃▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▂▃▃▃▃▄▄▆█████████████▇▆▆▄▄▃▂ ▄
  1.62 ms         Histogram: frequency by time        2.01 ms <

 Memory estimate: 1.45 KiB, allocs estimate: 28.
```

NEW:
```julia
julia> @benchmark CUDA.@sync permutedims($t, (3,2,1))
BenchmarkTools.Trial: 1661 samples with 1 evaluation.
 Range (min … max):  1.599 ms … 513.128 ms  ┊ GC (min … max): 0.00% … 0.58%
 Time  (median):     1.918 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.998 ms ±  22.740 ms  ┊ GC (mean ± σ):  0.18% ± 0.02%

                                              ▁▃██▇█▇▇▃        
  ▂▂▂▂▁▂▁▁▂▃▂▃▄▃▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▂▃▃▃▃▄▅▄▃▃▄▅██████████▇▇▄▄▄▃ ▃
  1.6 ms          Histogram: frequency by time        1.99 ms <

 Memory estimate: 1.39 KiB, allocs estimate: 26.
```

2. for high dimensional arrays, see: https://github.com/under-Peter/OMEinsum.jl/issues/133

maleadt commented on src/host/linalg.jl:
```diff
@@ -181,25 +181,30 @@ LinearAlgebra.lmul!(a::Number, B::AbstractGPUArray) = generic_lmul!(a, B)
 
 ## permutedims
 
+# TODO: implementation without the memory copy
```
You mean an in-place version?

maleadt commented:
Nice, thanks! Could you maybe add some comments on the approach? Makes it easier to understand when coming back to the code.

GiggleLiu commented:
I added some comments. The break down is because I added a new test, and it breaks for old Julia version.
This issue has been fixed by upstream, see #333 

I added a static condition expression to check the version.

maleadt commented:
Maybe now's the time to bump the Buildkite CI to 1.7 then; maybe do so in this PR.
